[[ Bug 21570 ]] Tree View Widget - select new elements automatically

bwmilby opened the pull request:
The proposed change is the add a new property to the tree view widget `hiliteNewElement` which when true will cause the widget to automatically highlight the added key.  That will yield 2 messaged - `dataChanged` and `hiliteChanged`.

bwmilby committed:
[[ Bug 21570 ]] Tree View Widget - select new elements automatically

The proposed change is the add a new property to the tree view widget `hiliteNewElement` which when true will cause the widget to automatically highlight the added key.  That will yield 2 messaged - `dataChanged` and `hiliteChanged`.

montegoulding commented:
@bwmilby I'm wondering about ensuring the highlighted node is visible if there is automatic highlighting

bwmilby commented:
> @bwmilby I'm wondering about ensuring the highlighted node is visible if there is automatic highlighting

If you `set the hilitedElement` to an node that is not visible, it currently does not scroll into view.  If you then scroll the view, it does properly show the highlight.  Adjusting the behavior should not be difficult, but I think it should be consistent in both cases.

montegoulding commented:
Hmm... what about a property to control it. `scrollHilitedElementIntoView` or something very camel casey like that ;-) It seems to me likely that most people would want it to scroll into view

bwmilby commented:
Sounds good.  I should be able to have something this weekend.

bwmilby committed:
[[ Tree View Widget ]] Add scrollHilitedElementIntoView property

The tree view widget now has the ability to scroll the selected row into
view. If `true`, this will happen when setting the **arrayData**,
setting the **hilitedElement**, and when adding a new row
(when **hiliteNewElement** is `true`).

One property has been added to achieve this:
* **scrollHilitedElementIntoView**: either `true` or `false`

The default value is `false` to match the behavior of previous versions
of the widget.

When selecting a row that is partially visible, the view will be
adjusted so that the full row is visible.

When changing the **readOnly** property, the view will only change if
the value is being set to `true` and the `Add new element` row is
currently visible.

bwmilby commented:
@montegoulding Here's a demo of how the new property impacts the widget.

![treeviewtest2](https://user-images.githubusercontent.com/29169921/45600103-c848bb00-b9bc-11e8-9aa4-afc1dfa9093d.gif)

![treeviewtest](https://user-images.githubusercontent.com/29169921/45599849-87e73e00-b9b8-11e8-8a19-d9c882220cd2.gif)

montegoulding commented:
This seems great @bwmilby !

montegoulding commented on extensions/widgets/treeview/notes/feature-scroll_into_view.md:
```diff
@@ -0,0 +1,19 @@
+# Properties
+
+The tree view widget now has the ability to scroll the selected row into 
```
@bwmilby I think you should discuss the new property `hiliteNewElement` here also.

bwmilby replied:
Added property to the note.

bwmilby committed:
[[ Tree View Widget ]] Add hiliteNewElement to feature note

montegoulding commented:
@livecode-vulcan review ok b7e4a49

livecode-vulcan commented:
💙 review by @montegoulding ok b7e4a4945574d4a4a697f9cb34235aec663be008

livecode-vulcan commented:
😞 test failure b7e4a4945574d4a4a697f9cb34235aec663be008

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/93)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/85)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/85)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/85)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/85)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1607)
* try-community-universal-ios-iphoneos11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.4/builds/134)
* try-community-universal-ios-iphonesimulator11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.4/builds/134)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1637)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1630)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1627)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1663)
* try-community-x86_64-win32: [failure](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1604)

montegoulding commented:
@livecode-vulcan retry b7e4a49

livecode-vulcan commented:
😎 test success b7e4a4945574d4a4a697f9cb34235aec663be008

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/94)
* try-community-armv7-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-sdk26_ndk16r15/builds/86)
* try-community-arm64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-sdk26_ndk16r15/builds/86)
* try-community-x86-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-sdk26_ndk16r15/builds/86)
* try-community-x86_64-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-sdk26_ndk16r15/builds/86)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/1608)
* try-community-universal-ios-iphoneos11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.4/builds/135)
* try-community-universal-ios-iphonesimulator11.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.4/builds/135)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/1638)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/1631)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/1628)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1664)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/1605)
